### DISCUSSION TITLE
What is the correct way to get the distance between and the corresponding closest points?

### DISCUSSION THREAD

**Discussion by BolunDai0216** [coal-library/coal#405] [Q&A] [answered]
Hi,

I just started to look into this library, and I noticed that when computing the distance between two objects

```python
import hppfcl
import numpy as np

box = hppfcl.Box(np.array([1.0, 1.0, 1.0]))
capsule = hppfcl.Capsule(1.0, 2.0)

M_box = hppfcl.Transform3f(np.eye(3), np.array([0, 0, 0]))
M_capsule = hppfcl.Transform3f(np.eye(3), np.array([10, 0, 0]))

req = hppfcl.DistanceRequest()
res = hppfcl.DistanceResult()

dis = hppfcl.distance(capsule, M_capsule, box, M_box, req, res)
```

It does not give any other information, like the corresponding closest points. Is possible to help me understand how I can obtain that information?

Thanks in advance!

**Comment by jcarpent** [comment]
Did you look at the `res` quantity? This is very similar to FCL here.

**Comment by BolunDai0216** [comment, marked as answer]
Thanks for the response!

So we can get the closest points using

```python
capsule_point = res.getNearestPoint1()
box_point = res.getNearestPoint2()
```

This seems to give the points in world frame coordinates, is there a way to directly get the points in body frame coordinates? 

Another question I have is, when I want to compute the distance between the same objects just with different `Transform3f`s, is there an `update` function for that?

**Comment by jcarpent** [comment]
> This seems to give the points in world frame coordinates, is there a way to directly get the points in body frame coordinates?

Just apply the inverse transformation on the points.

> Another question I have is, when I want to compute the distance between the same objects just with different `Transform3f`s, is there an `update` function for that?

Do a new call. You won't gain so much, except in specific scenarios. 
What do you expect with the `update`?

And please do not forget to cite this repository in your code and publications.

**Reply by BolunDai0216** [reply]
@jcarpent I was expecting it to be faster, but it seems like the current version is fast enough.

Another question I have is when I call `hppfcl.distance` is it returning a lower bound of the real distance? If that's the case, is there a way to get the real distance using `hppfcl`?

And I will cite this repo in my upcoming work!

**Reply by jcarpent** [reply]
Use the collision function instead. This will return a distance lower bound.

**Reply by BolunDai0216** [reply]
Should I also look at the `hppfcl.CollisionResult()` instance for the distance and closest points? When I look at it

```console
(Pdb) res_left
<hppfcl.hppfcl.CollisionResult object at 0x11e315810>
(Pdb) dir(res_left)
['__class__', '__delattr__', '__dict__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__le__', '__lt__', '__module__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', '__weakref__', 'addContact', 'cached_gjk_guess', 'cached_support_func_guess', 'clear', 'distance_lower_bound', 'getContact', 'getContacts', 'isCollision', 'numContacts', 'timings']
```

I did not see anything like `distance` or `getNearestPoint`.

**Reply by jcarpent** [reply]
`distance_lower_bound` is your distance info and `getContact(0)` will return the contact point if any as a `hppfcl.Contact`struct.

**Reply by jcarpent** [reply]
> @jcarpent I was expecting it to be faster, but it seems like the current version is fast enough.

Fast is a relative concept. What do you clearly mean here? Could you give real timings here?

**Reply by BolunDai0216** [reply]
What I meant was, if there exists an `update()`, then it would probably save time by not creating new variables at each time step. But using the `distance` function is fast enough for my purpose (less than 1 ms per call). I did not actually time it, but it was definitely faster than 1 ms.

**Reply by jcarpent** [reply]
Yes. You should be around the us scale
